byterun/shared_heap.c

```diff
@@ -220,6 +232,62 @@ static pool* pool_find(struct caml_heap_state* local, sizeclass sz) {
   r = local->avail_pools[sz];
   if (r) return r;
 
+/* Haven't managed to find a pool locally, try the global ones */
+  caml_plat_lock(&pool_freelist.lock);
+  if( pool_freelist.global_avail_pools[sz] ) {
+    r = pool_freelist.global_avail_pools[sz];
+
+    if( r ) {
+      pool_freelist.global_avail_pools[sz] = r->next;
+      r->next = 0;
+      local->avail_pools[sz] = r;
+
+      int free_objs = 0;
+      value* next_obj = r->next_obj;
+      while( next_obj ) {
+        free_objs++;
+        CAMLassert(next_obj[0] == 0);
+        next_obj = (value*)next_obj[1];
+      }
+
+      struct heap_stats tmp_stats = { 0 };
+
+      calc_pool_stats(r, sz, &tmp_stats);
+      caml_accum_heap_stats(&local->stats, &tmp_stats);
+      caml_remove_heap_stats(&pool_freelist.stats, &tmp_stats);
+
+      if (local->stats.pool_words > local->stats.pool_max_words)
+        local->stats.pool_max_words = local->stats.pool_words;    
+    }
+  }
+
+  /* There were no global avail pools, so let's adopt the full ones and try 
+     our luck sweeping them later on */
+  if( !r ) {
+    struct heap_stats tmp_stats = { 0 };
+
+    for(v = pool_freelist.global_full_pools[sz]; v; v = v->next ) {
+      calc_pool_stats(v, sz, &tmp_stats);
+    }
```

Comment:
I could move these pools somewhere else temporarily, unlock the platform lock, calculate their stats and _then_ move them to the local full pools.

Reply:
I think this is reasonable. Unless we find contention in practice, let's not complicate this code.